external-crates/move/crates/move-vm-runtime/src/tracing2/tracer.rs

```diff
@@ -0,0 +1,1812 @@
+#[cfg(feature = "gas-profiler")]
+mod vm_tracer_impl {
+    use crate::{
+        interpreter::{Frame, Interpreter},
+        loader::{Function, Loader},
+    };
+    use move_binary_format::{
+        errors::PartialVMError,
+        file_format::{ConstantPoolIndex, SignatureIndex},
+    };
+    use move_core_types::{
+        account_address::AccountAddress,
+        annotated_value::{MoveTypeLayout, MoveValue},
+        language_storage::TypeTag,
+    };
+    use move_trace_format::format::{
+        DataLoad, Effect as EF, Location, MoveTraceBuilder, Read, RefType, TraceIndex, TraceValue,
+        TypeTagWithRefs, Write,
+    };
+    use move_vm_types::{loaded_data::runtime_types::Type, values::Value};
+    use std::collections::BTreeMap;
+
+    /// Internal state for the tracer. This is where the actual tracing logic is implemented.
+    pub(crate) struct VMTracer<'a> {
+        trace: &'a mut MoveTraceBuilder,
+        link_context: Option<AccountAddress>,
+        pc: Option<u16>,
+        active_frames: BTreeMap<TraceIndex, FrameInfo>,
+        type_stack: Vec<RootedType>,
+        loaded_data: BTreeMap<TraceIndex, TraceValue>,
+        effects: Vec<EF>,
+    }
+
+    /// Information about a frame that we keep during trace building
+    #[derive(Debug, Clone)]
+    struct FrameInfo {
+        is_native: bool,
+        locals_types: Vec<LocalType>,
+        return_types: Vec<TagWithLayoutInfoOpt>,
+    }
+
+    /// A type tag, and the move type layout and reference information for that type if it is
+    /// computable without error. Due to runtime value depth restrictions you can have a valid type
+    /// whose type layout is not computable at runtime without error.
+    #[derive(Debug, Clone)]
+    struct TagWithLayoutInfoOpt {
+        tag: TypeTag,
+        layout: (Option<MoveTypeLayout>, Option<RefType>),
+    }
+
+    // Information about a function that we use for trace building
+    // All types are fully substituted
+    #[derive(Debug, Clone)]
+    struct FunctionTypeInfo {
+        ty_args: Vec<TypeTag>,
+        local_types: Vec<TagWithLayoutInfoOpt>,
+        return_types: Vec<TagWithLayoutInfoOpt>,
+    }
+
+    /// A runtime location can refer to the stack to make it easier to refer to values on the stack and
+    /// resolving them. However, the stack is not a valid location for a reference and all references
+    /// are rooted in a local so the Trace `Location` does not include the stack, and only `Local` and
+    /// `Indexed` locations.
+    #[derive(Debug, Clone)]
+    enum RuntimeLocation {
+        Stack(usize),
+        Local(TraceIndex, usize),
+        Indexed(Box<RuntimeLocation>, usize),
+        Global(TraceIndex),
+    }
+
+    /// The reference information for a local. This is used to track the state of a local in a frame.
+    /// * It can be a value, in which case the reference type is `Value`.
+    /// * It can be a local that does not currently hold a value (is "empty"), in which case the
```

Comment:
Great question! This is tracking the state of the local during the course of execution, so e.g., when we open a frame, there may be a local (of reference type) that starts out uninitialized, so this would show up as a local with a `LocalType` of `LocalType { layout: ..., ref_type: ReferenceType::Empty(RefType::Imm)`. This then gets filled later on throughout execution.